### PR TITLE
Add support for `select` op

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Select op takes a slice of a tensor at a given dimension. For example, for a 3D tensor, `t.select(1, 123)` means taking a slice along dim 1 at index 123, which is identical to `t[:, 123, :]`.

The reason why I write this PR is:
- `select` is a commonly used op in PyTorch, it is always good to have better op coverage.
- It is similar to `index_select` but easier, so this PR could serve as a reference for the `index_select`, and it is also a good place to discuss indexing about `index_select` here.

The current support of `select` is not complete, it has restrictions that require the input tensor to be a fusion input. And I don't think in `registry.cpp`, this PR is detecting all unsupported cases and rejecting them. But for now, I believe providing this level of support is sufficient.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/lower_index.cpp [hunk 1]
```diff
@@ -192,6 +192,14 @@ void IndexLowering::handle(const TernaryOp* top) {
   GpuLower::current()->propagateExprInfo(top, back());
 }
 
+void IndexLowering::handle(const SelectOp* sop) {
+  const auto input = lowerSrcIndex(sop->input(0), sop->output(0));
+  const auto out = lowerDstIndex(sop->output(0));
+  pushBack(
+      IrBuilder::create<SelectOp>(out, input, sop->getDim(), sop->input(1)));
+  GpuLower::current()->propagateExprInfo(sop, back());
+}
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/lower_index.cpp]: For `index_select`, I think we can do a `lowerSrcIndex` for the index tensor first, and then pass the lowered `TensorIndex` object to `getGlobalProducerStridedIndices` for the indexing of the given ID.

File: torch/csrc/jit/codegen/cuda/ir_utils.cpp [hunk 2]
```diff
@@ -767,6 +781,18 @@ std::vector<Expr*> getReductionOps(Fusion* fusion) {
   return red_ops;
 }
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/ir_utils.cpp:786]: TODO: the `getReductionOps` above needs clean up, will do it in a separate PR

File: torch/csrc/jit/codegen/cuda/arith.h [hunk 3]
```diff
@@ -143,6 +143,8 @@ TORCH_CUDA_CU_API WelfordResult WelfordRaw(
     // import IrBuilder just for this one interface.
     Int* init_N = nullptr);
 
+TORCH_CUDA_CU_API TensorView* select(TensorView* tv, int dim, Int* index);
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/arith.h:146]: Is it important to support `dim` as a symbolic value? (I hope not)

File: torch/csrc/jit/codegen/cuda/mutator.cpp [hunk 4]
```diff
@@ -137,6 +137,19 @@ void OptOutMutator::mutate(FullOp* fop) {
   IrBuilder::create<FullOp>(container, out, fill_value, fop->dtype());
 }
 
+void OptOutMutator::mutate(SelectOp* sop) {
+  Val* out = maybeMutated(sop->output(0));
+  Val* in = maybeMutated(sop->input(0));
+  Val* index = maybeMutated(sop->input(1));
+
+  if (out->sameAs(sop->output(0))) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/mutator.cpp]: Shouldn't we also do this check for `in` and `index`?

File: torch/csrc/jit/codegen/cuda/scheduler/utils.cpp [hunk 5]
```diff
@@ -972,7 +975,8 @@ std::vector<TensorView*> cacheInputs(Fusion* fusion, bool unroll) {
   // If we're going to unroll, make a cache of the inputs
   auto in_tvs = ir_utils::filterByType<TensorView>(fusion->inputs());
   for (auto tv : in_tvs) {
-    if (tv->uses().empty() || tv->isFusionOutput()) {
+    if (tv->uses().empty() || tv->isFusionOutput() ||
+        scheduler_utils::isSelectInput(tv)) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/scheduler/utils.cpp]: Please leave a comment about why tensors that are input to select are skipped. It should be something like: "Right now, tensors that are input to the select op can't be cached as they must be in global memory."

We should also add an error check in `TensorView::cacheAfter` to not cache select inputs.

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 6]
```diff
@@ -1615,7 +1612,10 @@ std::vector<Val*> Index::getGlobalProducerStridedIndices(
         loops,
         root_ind);
 
-    root_ind = getProducerIndexWithHalo(producer_tv, i, root_ind, consumer_tv);
+    if (root_dom[i] != selected_id) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp]: I think we should just assert this for now. `getProducerIndexWithHalo` should just return `root_ind` if the root domain is not extended with halo nor accessed through a shift op.

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 7]
```diff
@@ -1545,23 +1552,6 @@ std::vector<Val*> Index::getGlobalProducerStridedIndices(
...
-
-    TORCH_INTERNAL_ASSERT(
-        root_ind != nullptr,
-        "Couldn't find root mapping for ",
-        producer_tv->toString(),
-        " dim: ",
-        dim,
-        " id: ",
-        root_dom[dim]->toString());
-
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp:1564]: This is dead code, just cleaned it up

File: torch/csrc/jit/codegen/cuda/scheduler/registry.cpp [hunk 8]
```diff
@@ -1219,6 +1219,22 @@ class ReductionScheduler : public SchedulerEntry {
...
 
+    // Check that inputs of all select ops are fusion inputs
+    for (auto select : ir_utils::getSelectOps(fusion)) {
+      if (!select->input(0)->isFusionInput()) {
+        scheduler_debug_utils::canScheduleRejectReason(
+            ScheduleHeuristic::Reduction,
+            "Inputs of SelectOp must be fusion input.");
+        return false;
+      }
+      if (select->input(0)->uses().size() > 1) {
```
Comment (zasdfgbnm) [inline torch/csrc/jit/codegen/cuda/scheduler/registry.cpp:1230]: Maybe this restriction can be lifted, at least for the case where you have multiple select ops at the same dim.

**Review by naoyam — APPROVED** [review]
LGTM. Thanks for adding the support of `select`.

File: torch/csrc/jit/codegen/cuda/arith.h (hunk 3 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/arith.h:146]: I guess not. I don't see any use case for this.

File: torch/csrc/jit/codegen/cuda/lower_index.cpp (hunk 1 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_index.cpp]: I was wondering if it would make sense to allow an initial index map to be given to the index computation. In the case of `select`, we could initialize a map for the selected ID with the second argument of the select expression. For `index_select`, it would be a 2-step indexing like you mentioned. The first step would be indexing of the index argument, and then we could create a map of the selected ID to the lowered index tensor. Then `getGlobalProducerStridedIndices` would use this optional map to index into the selected root ID. This is similar to your approach, but maybe a little more generic?

File: torch/csrc/jit/codegen/cuda/mutator.cpp (hunk 4 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/mutator.cpp]: ~`in` and `index` are inputs to this op, so they are covered by `Expr::sameAs`~
Oh sorry, I thought you were commenting at `SelectOp::sameAs`. Yes, you are right, we should check it.

File: torch/csrc/jit/codegen/cuda/lower_index.cpp (hunk 1 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/lower_index.cpp]: I agree. Working on it.

File: torch/csrc/jit/codegen/cuda/mutator.cpp (hunk 4 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/mutator.cpp]: Is `Expr::sameAs` used here?

Looks like there's some inconsistency in the `mutate` functions. `mutate(UnaryOp*)` does `sameAs` both for its input and output. `mutate(BinaryOp*)` also looks at both input and output parameters, but it doesn't use `sameAs` but just looks at the pointer addresses.

Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/mutator.cpp]: > Is `Expr::sameAs` used here?

No, sorry for the confusion. I edited my previous comment: https://github.com/csarofeen/pytorch/pull/2179#discussion_r1020521096

Besides the inconsistency you mentioned, I also find a bug in the mutator of `ViewAsScalar`, I will write a separate PR to fix them.

File: torch/csrc/jit/codegen/cuda/lower_index.cpp (hunk 1 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/lower_index.cpp]: I did some relatively big refactor to implement it in this commit: https://github.com/csarofeen/pytorch/pull/2179/commits/87fee1daf1ae848a0e475998921bb9a259393966
@naoyam Could you please take another look to see if it makes sense to you?
That commit also resolves some of your review comments, but not all of them. I will work on some other review comments in later commits in this PR.

File: torch/csrc/jit/codegen/cuda/scheduler/utils.cpp (hunk 5 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/scheduler/utils.cpp]: added

File: torch/csrc/jit/codegen/cuda/index_compute.cpp (hunk 6 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp]: There is an assert inside `getProducerHaloOffset`:
```C++
  auto it = p2c.find(producer_id);
  // p2c should always have a mapping for producer_id. The only case
  // where no mapping exists for a producer axis is when it is a
  // reduction axis. Since this function is only used for indexing
  // producer tensors, where reduction axes are skipped, producer_id
  // should never be a reduction axis.
  TORCH_INTERNAL_ASSERT(it != p2c.end());
```
which fails for the select id. Do we want to remove this assert and return `0` if not mapped?

File: torch/csrc/jit/codegen/cuda/lower_index.h [hunk 9]
```diff
@@ -65,7 +66,10 @@ class TORCH_CUDA_CU_API IndexLowering : private OptOutConstDispatch {
 
   void generate(const std::vector<Expr*>& exprs);
 
-  Val* lowerSrcIndex(Val* val, Val* dst) const;
+  Val* lowerSrcIndex(
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_index.h:74]: Does the source IDs of the map need to be concrete? Please add a comment about it.

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 10]
```diff
@@ -1591,18 +1575,27 @@ std::vector<Val*> Index::getGlobalProducerStridedIndices(
       continue;
     }
 
-    TORCH_INTERNAL_ASSERT(
+    Val* root_ind = nullptr;
+    auto override_it = override_index.find(root_dom[i]);
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp:1581]: I was thinking about passing the optional map to `getTensorIndexFromIdGraph` to provide an initial ID-to-index map. That would be more consistent if we would want to allow the same initial map in consumer indexing.

That said, I think this is good enough for now given that the whole indexing code would be redesigned. 

Pinging @csarofeen

File: torch/csrc/jit/codegen/cuda/index_compute.cpp (hunk 6 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp]: Yeah, I think we should remove the assert, but I'd feel more assured with some validation that the producer ID is a selected ID.

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 11]
```diff
@@ -1615,7 +1608,10 @@ std::vector<Val*> Index::getGlobalProducerStridedIndices(
         loops,
         root_ind);
 
-    root_ind = getProducerIndexWithHalo(producer_tv, i, root_ind, consumer_tv);
+    if (!override_index.count(root_dom[i])) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp]: Added a comment the original comment. I think we should remove the assertion and also this `if`. It looks inconsistent to have a predicate only for `getProducerIndexWithHalo`, whereas `getProducerIndexWithGather`, etc. are just called always.

**Review by naoyam — APPROVED** [review]
LGTM. Added some minor comments.

File: torch/csrc/jit/codegen/cuda/index_compute.cpp (hunk 11 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp]: changed

File: torch/csrc/jit/codegen/cuda/lower_index.h (hunk 9 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/lower_index.h:74]: added

File: torch/csrc/jit/codegen/cuda/index_compute.cpp (hunk 6 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp]: I moved the check inside `getProducerIndexWithHalo`, something like
```C++
if (override) {
  offset = 0;
} else {
  assert 
}
```

**Comment by csarofeen** [comment]
Is select an alias operation?

**Comment by zasdfgbnm** [comment]
> Is select an alias operation?

It is, it returns a view of the input tensor.

File: torch/csrc/jit/codegen/cuda/index_compute.cpp (hunk 10 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp:1581]: Agreed

**Comment by naoyam** [comment]
> > Is select an alias operation?
> 
> It is, it returns a view of the input tensor.

In nvFuser, even a view isn't really an alias as it creates a new (transformed) copy. I don't think we have a true alias in nvFuser. Is it different in PyTorch? Is a view a true "view" to another tensor?

**Comment by zasdfgbnm** [comment]
> > > Is select an alias operation?
> > 
> > 
> > It is, it returns a view of the input tensor.
> 
> In nvFuser, even a view isn't really an alias as it creates a new (transformed) copy. I don't think we have a true alias in nvFuser. Is it different in PyTorch? Is a view a true "view" to another tensor?

Yes, in eager mode, `view` is a true view. If the stride is not compatible with the viewed shape, then `view` will throw an error. For this case, user can use `reshape` to make a copy of the tensor with compatible stride.

**Comment by naoyam** [comment]
> > > > Is select an alias operation?
> > > 
> > > 
> > > It is, it returns a view of the input tensor.
> > 
> > 
> > In nvFuser, even a view isn't really an alias as it creates a new (transformed) copy. I don't think we have a true alias in nvFuser. Is it different in PyTorch? Is a view a true "view" to another tensor?
> 
> Yes, in eager mode, `view` is a true view. If the stride is not compatible with the viewed shape, then `view` will throw an error. For this case, user can use `reshape` to make a copy of the tensor with compatible stride.

Hmm, so I'm assuming we somehow handle the semantic difference, right? I wonder what would happen when writing to view tensors.

**Comment by zasdfgbnm** [comment]
> > > > > Is select an alias operation?
> > > > 
> > > > 
> > > > It is, it returns a view of the input tensor.
> > > 
> > > 
> > > In nvFuser, even a view isn't really an alias as it creates a new (transformed) copy. I don't think we have a true alias in nvFuser. Is it different in PyTorch? Is a view a true "view" to another tensor?
> > 
> > 
> > Yes, in eager mode, `view` is a true view. If the stride is not compatible with the viewed shape, then `view` will throw an error. For this case, user can use `reshape` to make a copy of the tensor with compatible stride.
> 
> Hmm, so I'm assuming we somehow handle the semantic difference, right? I wonder what would happen when writing to view tensors.

I don't know. I remember seeing something called `view_copy`, not sure how it works though. Probably TorchScript might convert the IR to SSA before feeding to nvfuser? @jjsjann123

**Comment by csarofeen** [comment]
Yeah, this was why I asked. We would need to figure out how to support the alias analysis of the op to pipe it through.

**Comment by jjsjann123** [comment]
sounds like we needed to have an alias analysis to ensure that changing the `select` op to `select_copy` preserves correct semantics. We did that for other `view` like ops.
